### PR TITLE
Review hugo july 2024

### PR DESCRIPTION
**Pull request by Bisaloo**
Todo:

- [x] Files in https://github.com/TRACE-LAC/sivirep_incidence will need to be converted from rda to rds

Extra review comments:

- `geo_filtra(data_limpia)` with the other args `NULL` returns an empty data.frame. I would expect it to return `data_limpia` instead
- can the epitrix dependency be replaced by cleanepi or another data cleaning package?
- Should `longitud` in `concatenar_vals_token()` be a parameter or computed from `length(vals)`?
- Missing arg in `import_pob_incidencia()` call: https://github.com/epiverse-trace/sivirep/blob/93799b86acdef5ed3d27e7ad9a899831c01745c7/R/checking_data.R#L1029
- Missing arg in `validar_area_geo()` calls:  https://github.com/epiverse-trace/sivirep/blob/93799b86acdef5ed3d27e7ad9a899831c01745c7/R/checking_data.R#L542 & https://github.com/epiverse-trace/sivirep/blob/93799b86acdef5ed3d27e7ad9a899831c01745c7/R/checking_data.R#L589
- The example in `limpiar_val_atipic()` is not demonstrating what this function is doing
- Should `col_med_uni` be set to 1 in `convert_edad()` to avoid accidentally converting the age again when it's already be converted to years?

Blockers for the CRAN release:

- CRAN policies don't allow writing to the user computer without their explicit consent. In other words, the `cache` argument needs to let the user opt-in and let them choose the location. Since R 4.0.0, you could use the `tools::R_user_dir()` function to help you: https://search.r-project.org/R/refmans/tools/html/userdir.html. The ColOpenData devs have had the same issue so it may be worthwhile to share strategies here.

After the first CRAN release, it would be good to:

- Harmonize patterns and function usage in the codebase. In particular, harmonize base vs tidy:
  - `grep()` vs `str_detect()`
  - `x[order()]` vs `arrange()`

### CODE REVIEW

**Comment by GeraldineGomez** [comment]
Hi @Bisaloo,

Thank you very much for your review, contribution, and comments 🤗🚀🙏🩷.

- [x] The incidence population files have been converted to RDS files.

Regarding the epixtrix aspect, sivirep currently uses one function from it. Perhaps it would be better to implement this function directly in the code to avoid adding a dependency. What do you think?

I am still reviewing the PR and the additional comments.

File: R/import_data.R [hunk 1]
```diff
@@ -212,30 +208,28 @@ import_sep_data <- function(ruta_data = NULL, cache = TRUE) {
...
+  ruta_archivo <- file.path(ruta_extdata, nomb_archivo)
+  if (!file.exists(ruta_archivo) || !cache) {
+    respuesta_archivo <- realizar_peticion_http(ruta_data)
+    if (httr2::resp_status(respuesta_archivo) == 200) {
+      conten_archivo <- httr2::resp_body_raw(respuesta_archivo)
+      con_archivo <- file(ruta_archivo, "wb")
+      on.exit(close(con_archivo))
+      if (length(conten_archivo) > 0) {
+        writeBin(conten_archivo, con_archivo)
       }
```
Comment (GeraldineGomez) [inline R/import_data.R:227]:
```suggestion
      if (length(conten_archivo) > 0) {
        writeBin(conten_archivo, con_archivo)
      }
      on.exit(close(con_archivo))
```

**Review by GeraldineGomez — COMMENTED** [review]
Hi @Bisaloo,

Many thanks for reviewing the package 🤗🚀🙏🩷. Your changes significantly improved the code's cleanliness, and streamlined several steps. Thanks for enhancing the overall quality of the package.

While testing the changes with the report template, I encountered some errors. To address these, I created a new branch [fix-review-items](https://github.com/epiverse-trace/sivirep/tree/fix-review-items) to resolve the issues and to implement additional fixes based on the mini user testing I conducted last week and the documentation review. In terms of functionality, the features in this branch are working well, and the unit tests are passing. I also wanted to avoid generating conflicts in case you were making additional changes. Please let me know if you agree with this strategy or if you prefer another approach to include these fixes. 

I tried to fix the problem with spelling because the es_ES dictionary does not exist by default in the operating systems. I included some commands that work on Ubuntu, but it was more difficult for other operating systems. For Windows and macOS, the issue is still unresolved.

Additionally, I have been working on the `import_data_event_function`, designing the best strategy to include user agreement and using the `R_user_dir`. However, after testing the function several times with different values, the cache directory returned errors, and the file did not close correctly. As a result, I am still using the `extdata` folder but have added two new parameters. The first one, `ruta_dir`, allows the user to specify the folder to store the data. The second one, `consentimiento`, determines if the user wants to store the data in case the folder is not specified. If the folder is not specified and consent is given, the data will be stored in the `extdata` folder. This logic depends on the `cache` parameter. The new implementation of the function is also in the branch [fix-review-items](https://github.com/epiverse-trace/sivirep/tree/fix-review-items). Please let me know what you think about this implementation.

On another note, Zulma is wondering if you are planning to add more changes to the review or if we can consider this version stable and proceed with submitting it to CRAN?

**Comment by GeraldineGomez** [comment]
I got the following errors:

image

image

I have a similar issue with `extdata` folder so I realized that the issue was related to handling the file, specifically, there was a problem with closing the file properly.

To resolve this, I modified the code from:

```{r}
on.exit(close(con_archivo))
if (length(conten_archivo) > 0) {
         writeBin(conten_archivo, con_archivo)
}
```
to 

```{r}
if (length(conten_archivo) > 0) {
         writeBin(conten_archivo, con_archivo)
}
close(con_archivo)
```
I tested the code again using R_user_dir, and it worked well, confirming that the problem was indeed with the `on.exit `mechanism. It appears that when using `on.exit(close(con_archivo))`, I checked, and the process was still holding onto the file, preventing it from being closed properly.

I was wondering if it is necessary to update the supported R version in the DESCRIPTION file, as the R_user_dir function is only available in R versions >= 4.0.0 or we could implement something like the following code to handle 3.5.0 and use $HOME, as you mentioned:

```{r}
  base_dir <- Sys.getenv("R_USER") 
  if (base_dir == "") {
    base_dir <- file.path(Sys.getenv("HOME"), ".local/share/R")
  }
  ruta_dir <- file.path(base_dir, "cache", "sivirep")
```

**Comment by Bisaloo** [comment]
> I tested the code again using R_user_dir, and it worked well, confirming that the problem was indeed with the on.exit mechanism. It appears that when using on.exit(close(con_archivo)), I checked, and the process was still holding onto the file, preventing it from being closed properly.

Okay, let's go with this then. I may try to dig more later as this is unexpected (to me at least) behaviour so I'd like to understand what's happening. Thanks for debugging!

> I was wondering if it is necessary to update the supported R version in the DESCRIPTION file, as the R_user_dir function is only available in R versions >= 4.0.0 or we could implement something like the following code to handle 3.5.0 and use $HOME, as you mentioned:

I think it's better to start depend on R 4.0.0. Note that you already have this dependency indirectly via dplyr and others as the tidyverse policy is to support only the last 5 minor versions, which means they started using 4.0.0 as their minimum version a couple of months ago.

It has now been released a while ago so I think it's quite safe and it's a good practice to simplify our internal code to facilitate maintenance.

**Comment by GeraldineGomez** [comment]
Yes, I agree. The `on.exit` mechanism is typically used to prevent issues like this.

It seems appropriate to officially set R 4.0.0 as our minimum supported version. I'll go ahead and update the DESCRIPTION file.

Thanks, Hugo!

**Pull request merged**